src/node_url.h

```diff
@@ -529,6 +529,91 @@ struct url_host {
   url_host_value value;
   enum url_host_type type;
 };
+
+class URL {
+ public:
+  static void Parse(const char* input,
+                    const size_t len,
+                    enum url_parse_state state_override,
+                    struct url_data* url,
+                    struct url_data* base,
+                    bool has_base);
+
+  URL(const char* input, const size_t len) {
+    Parse(input, len, kUnknownState, &context_, nullptr, false);
+  }
+
+  URL(const char* input, const size_t len, URL* base) {
+    if (base != nullptr)
+      Parse(input, len, kUnknownState, &context_, &(base->context_), true);
+    else
+      Parse(input, len, kUnknownState, &context_, nullptr, false);
+  }
+
+  URL(const char* input, const size_t len,
+      const char* base, const size_t baselen) {
+    if (base != nullptr && baselen > 0) {
+      URL _base(base, baselen);
+      Parse(input, len, kUnknownState, &context_, &(_base.context_), true);
+    } else {
+      Parse(input, len, kUnknownState, &context_, nullptr, false);
+    }
+  }
+
+  explicit URL(std::string input) :
+      URL(input.c_str(), input.length()) {}
+
+  URL(std::string input, URL* base) :
+      URL(input.c_str(), input.length(), base) {}
+
+  URL(std::string input, std::string base) :
+      URL(input.c_str(), input.length(), base.c_str(), base.length()) {}
+
+  int32_t flags() {
+    return context_.flags;
+  }
+
+  int port() {
+    return context_.port;
+  }
+
+  std::string protocol() {
+    return context_.scheme;
+  }
```

Comment:
I think there’s no reason to not make these methods return const references, i.e. `const std::string& protocol() const { … }` (except `path()`)?